analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs

```diff
@@ -236,21 +236,23 @@ private IEnumerable<ExplodedNode> ExceptionSuccessors(ExplodedNode node, Excepti
         state = node.Block.EnclosingRegion(ControlFlowRegionKind.Catch) is not null
             ? state.PushException(exception)        // If we're nested inside catch, we need to preserve the exception chain
             : state.SetException(exception);        // Otherwise we track only the current exception to avoid explosion of states after each statement
-        while (tryRegion is not null)
+
+        var handlersAreExhaustive = false;
+        while (tryRegion is not null && !handlersAreExhaustive)
         {
             var reachableHandlers = tryRegion.EnclosingRegion.NestedRegions.Where(x => x.Kind != ControlFlowRegionKind.Try && IsReachable(x, state.Exception)).ToArray();
             foreach (var handler in reachableHandlers)  // CatchRegion, FinallyRegion or FilterAndHandlerRegion
             {
+                handlersAreExhaustive |= CatchesAll(handler) || state.Exception.Type.DerivesFrom(handler.ExceptionType);
                 yield return new(cfg.Blocks[handler.FirstBlockOrdinal], state, null);
             }
-            if (reachableHandlers.Length != 0)
-            {
-                yield break;
-            }
             tryRegion = tryRegion.EnclosingRegion.EnclosingRegionOrSelf(ControlFlowRegionKind.Try); // Inner catch for specific exception doesn't match. Go to outer one.
         }
 
-        yield return new(cfg.ExitBlock, node.State.SetException(exception), null);  // catch without finally or uncaught exception type
+        if (!handlersAreExhaustive)
```

Comment:
Do we have UT for this specific scenario?

Reply:
Yes, [this one](https://github.com/SonarSource/sonar-dotnet/blob/cfee18a155ab7bf547d822d6a414cc9e45eebbbd/analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Throw.cs#L698-L740) checks for this specifically. And there are a lot of tests testing it indirectly.